HIP/hip-435.md

```diff
@@ -0,0 +1,131 @@
+needs-council-approval: Yes
+status: 
+last-call-date-time: 
+created: 2022-04-13
+discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/436
+updated: 
+requires: 
+---
+
+## Abstract
+
+Updates the record file definition to be done via a protobuf.
+
+Introduces Sidecar records - a generic way for Hedera nodes to externalise additional information about transactions that is not part of TransactionRecords.
+
+Includes BlockNumber changes for the record files.
+
+## Motivation
+
+Using a protobuf definition for the record file schema is far preferable to having a custom parser or random specification someplace (currently a word doc).
+Externalising verbose information about transaction executions without bloating the main TransactionRecord object with potentially unnecessary information and requiring mirror nodes to always download and process that information.
+
+## Rationale
+
+There are use-cases requiring a detailed and verbose output for a HAPI transaction execution. Users of Hedera should have the option to query a mirror node for this information. Meanwhile, bloating the TransactionRecords with information that would be unnecessary in most cases is not an ideal option. It is a much better design decision to create sidecar records that can be downloaded and processed only by interested parties.
+
+**Design Goal #1**
+Mirror nodes should not be required to download and process verbose transaction information. The sidecar transaction records will be externalised in separate record files.
+
+**Design Goal #2**
+One transaction can have 0, 1 or more than 1 sidecar records related to it.
+
+**Design Goal #3**
+It should be possible and relatively easy to add new types of sidecar transaction records.
+
+## User stories
+
+As a Hedera user, I want the to have the ability to access verbose information about transactions.
+
+As an operator of a mirror node, I want the ability to download and process only the most important record-related information.
+
+As an operator of a mirror node, I want the ability to download all the information about records execution that a Hedera node can provide.
+
+## Specification
+
+Introduce a sidecar transaction record protobuf message and a sidecar record file protobuf message to enable tracking of additional information about records execution.
+
+### TransactionSideCarRecord
+
+`TransactionSideCarRecord` is a protobuf which will be used to create sidecar records complementing `TransactionRecord` and storing additional information about a transaction execution.
+The `TransactionSideCarRecord`s created by Hedera nodes will be externalised in a new file, part of the new RecordStream structure described bellow.
+
+```
+message TransactionSideCarRecord {
+    // Consensus timestamp will be the same as the consensus timestamp of the 
+    // transaction the side car is related to. This offers a convenient 
+    // way to match record to sidecar. 
+    Timestamp consensusTimestamp = 1;
```

Comment:
So this is not a one-to-one index, correct?  We can have a sidecar message for storage change and then another sidecar message for contract actions?

Reply:
Yes, and both sidecar messages will have the same timestamp since they're related to the same transaction record.